### PR TITLE
docs: RFC of enclosing column id

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Propose to enclose the usage of `RegionId` into the region engine only.

[Rendered](https://github.com/GreptimeTeam/greptimedb/pull/2983/files?short_path=aeafd64#diff-aeafd64185a5f1ab4114f0430ce1635fbaaab400f3e7dc2a3327a824955ec37a)

Issue: https://github.com/GreptimeTeam/greptimedb/issues/2982

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: docs/rfcs/2023-12-22-enclose-column-id.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+Author: "Ruihang Xia <waynestxia@gmail.com>"
+---
+
+# Summary
+This RFC proposes to enclose the usage of `RegionId` into the region engine only.
+
+# Motivation
+`ColumnId` is an identifier for columns. It's assigned by meta server, stored in `TableInfo` and `RegionMetadata` and used in region engine to distinguish columns.
+
+At present, Both Frontend, datanode and Metasrv are aware of `ColumnId` but it's only used in region engine. Thus this RFC proposes to remove it from Frontend (mainly used in `TableInfo`) and Metasrv.
```
Comment (WenyXu) [inline docs/rfcs/2023-12-22-enclose-column-id.md:14]: nit:
```suggestion
At present, Both Frontend, Datanode and Metasrv are aware of `ColumnId` but it's only used in region engine. Thus this RFC proposes to remove it from Frontend (mainly used in `TableInfo`) and Metasrv.
```

**Review by WenyXu — COMMENTED** [review]
> A column is identified by its name in all components.

What will happen If the user removes a field and adds the removed field back to the table later?

IMO, both the data and indexes of the newly added field should be empty.

**Comment by waynexia** [comment]
> > A column is identified by its name in all components.
> 
> What will happen If the user removes a field and adds the removed field back to the table later?
> 
> IMO, both the data and indexes of the newly added field should be empty.

It should have the same behavior as the present. The first `ALTER DROP` will remove the column, and `ALTER ADD` will add a new column with a different column id (inside the region engine).

**Comment by WenyXu** [comment]
> It should have the same behavior as the present. The first ALTER DROP will remove the column, and ALTER ADD will add a new column with a different column id (inside the region engine).

I see. It seems like we have a duplicate column ID.

File: docs/rfcs/2023-12-22-enclose-column-id.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+# Motivation
+`ColumnId` is an identifier for columns. It's assigned by meta server, stored in `TableInfo` and `RegionMetadata` and used in region engine to distinguish columns.
+
+At present, Both Frontend, Datanode and Metasrv are aware of `ColumnId` but it's only used in region engine. Thus this RFC proposes to remove it from Frontend (mainly used in `TableInfo`) and Metasrv.
+
+# Details
+
+`RegionId` is used widely on both read and write paths. Removing it from Frontend and Metasrv implies several things:
+
+- A column may have different column id in different regions.
```
Comment (evenyag) [inline docs/rfcs/2023-12-22-enclose-column-id.md:20]: For simplicity, we can still assume columns have the same column ids across regions. The procedure maintains consistency between regions and the region engine should ensure alterations are idempotent.

File: docs/rfcs/2023-12-22-enclose-column-id.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
+---
+Feature Name: Enclose Region Id
```
Comment (evenyag) [inline docs/rfcs/2023-12-22-enclose-column-id.md:2]:
```suggestion
Feature Name: Enclose Column Id
```

**Review by evenyag — APPROVED** [review]
I suppose they should be Column Ids. 😓 

Anyway, I agree that only region engines need to use column id.

**Comment by MichaelScofield** [comment]
We already have some serious customers, is it possible to achieve your goal without breaking changes? Like "deprecating" the column id. Exporting then importing the tables require a not neglectable time of write suspension.

File: docs/rfcs/2023-12-22-enclose-column-id.md (hunk 2 above)
Comment (waynexia) [reply docs/rfcs/2023-12-22-enclose-column-id.md:20]: Thanks for your kind review. I've added this paragraph to the explanation.

**Comment by waynexia** [comment]
> We already have some serious customers, is it possible to achieve your goal without breaking changes? Like "deprecating" the column id. Exporting then importing the tables require a not neglectable time of write suspension.

Don't worry, this won't be included in 0.5

**Review by WenyXu — APPROVED** [review]

**Pull request merged**